README
This is acer_acpi version 0.8.2

### THIS IS VERY MUCH A WORK IN PROGRESS.  NO WARRANTY GIVEN.  IF YOU
### BREAK YOUR LAPTOP USING THIS DRIVER, YOU GET TO KEEP THE PIECES AND
### I WILL ACCEPT ABSOLUTELY NO RESPONSIBILITY.

acer_acpi is a driver to allow you to control various aspects of your Acer
laptop hardware using the proc filesystem.  It was originally developed in
order to activate the wireless LAN card under a 64-bit version of Linux,
as acerhk[1] (the previous fine solution to the problem) relied on making BIOS
calls which are not allowed from a 64-bit OS.

Please read the INSTALL file for installation instructions.

PLEASE NOTE: I (Mark Smith) AM NO LONGER ACTIVELY DEVELOPING ACER_ACPI.
(many would note that I never actively developed it! :-)

acer_acpi is now being maintained by Carlos Corbacho <cathectic@gmail.com>

[1] acerhk: http://www.cakey.de/acerhk/

Usage
*****

As root, modprobe acer_acpi

(To enable debugging, add debug=X, where X is between 0 and 2).

acer_acpi creates entries under /proc/acpi/acer and
/sys/devices/platform/acer_acpi, which enables you to control some of the
following (varies between models):

* the wireless LAN card
* inbuilt Bluetooth adapter
* mail LED of your laptop
* brightness of the LCD panel

The syntax for each is the same:

PLEASE NOTE: "enabled: X" support has now been removed

To read the status of a device (0=off, 1=on):
cat /proc/acpi/acer/{device}
or
cat /sys/devices/platform/acer_acpi/{device}

To enable a feature:
echo 1 > /proc/acpi/acer/{feature}
or
echo 1 > /sys/devices/platform/acer_acpi/{device}

To disable a feature:
echo 0 > /proc/acpi/acer/{feature}
echo 0 > /sys/devices/platform/acer_acpi/{device}

(brightness and mail LED are not visible in sysfs/ platform device - these are
properly registered with the relevant kernel subsystem).

To enable a feature on loading the module:
modprobe acer_acpi {feature}=1

To disable a feature on loading the module (default):
modprobe acer_acpi {feature}=0

Where {feature} is either wireless, bluetooth, 3G, or mailled. (You can have 
multiple {feature}'s specified on the modprobe line).

For laptops with a "brightness" option, you can write a value between 0 to 15 
(or 0x0 to 0xF).

Notes
*****

The behaviour of the wireless LED will depend on your hardware and driver
combination.

e.g. With the BCM4318 on the 5020 series:

ndiswrapper: Light blinks on when transmitting
bcm43xx/ 
bcm43xx_mac80211: Solid light, blinks off when transmitting

Also, acer_acpi does not yet have any means of deciding whether to
enable or disable entries in /proc/acpi/acer - this means that for
hardware that isn't there, you can still read/ write the corresponding proc
file.

So: Just because it _says_ "bluetooth" in /proc/acpi/acer, doesn't mean 
you now actually have a Bluetooth device.

Credits
*******

Olaf Tauber, who did the real hard work when he developed acerhk
http://www.informatik.hu-berlin.de/~tauber/acerhk
All the authors of laptop acpi modules in the kernel, whose work
was an inspiration and a source of good code, particularly John Belmonte,
author of the toshiba_acpi driver.
Mathieu Segaud, who solved the problem with having to modprobe the driver
twice in version 0.2 and below.
Jim Ramsay, who added support for the new WMID interface

Contact
*******

If you have problems with the driver, please file a bug on the website and
include the following information:
1. name of your laptop (e.g. Acer Aspire 5021)
2. contents of your ACPI DSDT (cat /proc/acpi/dsdt).  I can disassemble
   it if necessary, so I'll accept the binaries.

Please check http://code.google.com/p/aceracpi/wiki/SupportedHardware first to
see if your hardware is supported or unsupported.

The latest version of acer_acpi can be found here:

http://code.google.com/p/aceracpi


NEWS
2007-09-03 Version 0.8.2
	Really fix backlight code on older kernels
	Fix write errors with /proc/acpi/acer/brightness
	Fix problem with driver registration on WMID

2007-09-02 Version 0.8.1
	Fix backlight code on older (pre 2.6.21) kernels

2007-09-01 Version 0.8
	Register backlight with backlight driver
	Restored 3G support
	Added new sysfs platform driver to /sys/devices/platform/acer_acpi
	Added EC support for Aspire 3020, 3040, 5040, 5650 and 9300
	Code clean up - switched C99 types to kernel types, removed typedefs

2007-08-02 Version 0.7
	Dropped 'enabled: X' support
	Added limited direct EC support for Aspire 5020 and TravelMate 2490
	Fixed broken AMW0 values on hibernate
        Added new sysfs interface to /sys/devices/platform/acer_acpi

2007-07-17 Version 0.6.3
	Fix Bash 3.2 bug in Makefile (really fixes *buntu build bug)

2007-07-15 Version 0.6.2
	Fix packaging bug from 0.6.1 (directory name accidentally
	dropped acer_acpi-)

2007-07-13 Version 0.6.1
	Remove 2.4 kernel check, as it breaks *buntu systems

2007-07-06 Version 0.6
	Cleaned up code
	Added WMID support
	Makefile no longer gives harmless errors on older 2.6 kernels

2007-05-05 Version 0.5
	New maintainer added: Carlos Corbacho <cathectic@gmail.com>	
	cat /proc/acpi/acer/{feature} now works - see ChangeLog for more

2007-03-31 Version 0.4
	Compiles under later 2.6 series kernels.
	New home at code.google.com

2005-08-30 Version 0.3
	Fixed a bug in the call to acpi_bus_register_driver, so the module
	no longer needs loading twice.

2005-08-28 Version 0.2
	First public release, this version actually does some useful things.

2005-08-27 Version 0.1
	Initial release that enables wireless on modprobe.  And that's it.


acer_ec/acer_ec.pl
#!/usr/bin/perl -w

#Copyright (C) 2007  Michael Kurz     michi.kurz (at) googlemail.com
#Copyright (C) 2007  Petr Tomasek     tomasek (#) etf,cuni,cz
#Copyright (C) 2007  Carlos Corbacho  cathectic (at) gmail.com
#
#Version 0.4 (2007-09-04)
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 3
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


require 5.004;

use strict;
use Fcntl;
use POSIX;
use File::Basename;

sub initialize_ioports
{
  sysopen (IOPORTS, "/dev/port", O_RDWR)
    or die "/dev/port: $!\n";
  binmode IOPORTS;
}

sub close_ioports
{
  close (IOPORTS)
    or print "Warning: $!\n";
}



sub inb
{
  my ($res,$nrchars);
  sysseek IOPORTS, $_[0], 0 or return -1;
  $nrchars = sysread IOPORTS, $res, 1;
  return -1 if not defined $nrchars or $nrchars != 1;
  $res = unpack "C",$res ;
  return $res;
}

# $_[0]: value to write
# $_[1]: port to write
# Returns: -1 on failure, 0 on success.
sub outb
{
  if ($_[0] > 0xff)
  {
    my ($package, $filename, $line, $sub) = caller(1);
    print "\n*** Called outb with value=$_[1] from line $line\n",
          "*** (in $sub). PLEASE REPORT!\n",
          "*** Terminating.\n";
    exit(-1);
  }
  my $towrite = pack "C", $_[0];
  sysseek IOPORTS, $_[1], 0 or return -1;
  my $nrchars = syswrite IOPORTS, $towrite, 1;
  return -1 if not defined $nrchars or $nrchars != 1;
  return 0;
}

sub wait_write
{
	my $i = 0;
	while ((inb(0x66) & 0x02) && ($i < 10000)) {
		sleep(0.01);
		$i++;
	}
	return -($i == 10000);
}

sub wait_read
{
	my $i = 0;
	while (!(inb(0x66) & 0x01) && ($i < 10000)) {
		sleep(0.01);
		$i++;
	}
	return -($i == 10000);
}

sub send_ec
{
	if (!wait_write()) { outb($_[0], 0x66); }
	if (!wait_write()) { outb($_[1], 0x62); }
}

sub write_ec
{
	if (!wait_write()) { outb(0x81, 0x66 ); }
	if (!wait_write()) { outb($_[0], 0x62); }
	if (!wait_write()) { outb($_[1], 0x62); }
}

sub read_ec
{
	if (!wait_write()) { outb(0x80, 0x66 ); }
	if (!wait_write()) { outb($_[0], 0x62); }
	if (!wait_read())  { inb(0x62); }
}

sub print_regs
{
	initialize_ioports();

	my @arr = ("00","10","20","30","40","50","60","70","80","90","A0","B0","C0","D0","E0","F0", "");

	my $i = 0;
	my $t = 0;
	print "\n  \t00\t01\t02\t03\t04\t05\t06\t07\t|\t08\t09\t0A\t0B\t0C\t0D\t0E\t0F\n";
	print "  \t__\t__\t__\t__\t__\t__\t__\t__\t|\t__\t__\t__\t__\t__\t__\t__\t__\n";
	print "00 |\t";
	for ($i = 0; $i < 256; $i++)
	{
		$t = read_ec($i);
		print $t;
		print "\t";
		if ((($i + 1) % 8) == 0){
			if ((($i + 1) % 16) == 0) {
				if ($i != 255) { print "\n$arr[(($i-(($i + 1) % 16)) / 16) + 1] |\t"; }
			} else {
				print "|\t";
			}
		}
	}
	
	print "\n";
	
	close_ioports();
}

sub write_temp
{
	initialize_ioports();
	write_ec($_[0],$_[1]);
	close_ioports();
}

sub testnum
{
	my $i;
	for ($i = 0; $i<256;$i++) {
		if ($_[0] eq "$i") { return 1 };
	}
	return 0;
}

my $ii;

if (!$ARGV[0]){
        print "wrong arguments!\n";
	print "usage:\n";
	print "\'acer_ec regs\' \t\t\t\tdumps all ec registers\n";
	print "\'acer_ec ledon\' \t\t\t\tswitch on 'mail LED' (WMID)\n";
	print "\'acer_ec ledoff\' \t\t\t\tswitch off 'mail LED' (WMID)\n";
	print "\'acer_ec getled\' \t\t\t\tget 'mail LED' status (WMID)\n";
	print "\'acer_ec getled2\' \t\t\t\tget 'mail LED' status(AMW0)\n";
	print "\'acer_ec getwireless\' \t\t\t\tget 'wireless' status (AMW0)\n";
	print "\'acer_ec gettouch\' \t\t\t\tis the touchpad disabled?\n";
	print "\'acer_ec setfanthresh <temp>\' \t\t\t\tset temperature threshhold to <temp>, DANGEROUS!\n";
	print "\'acer_ec getfanthresh\' \t\t\t\tget temperature threshhold\n";
	print "\'acer_ec <temp-number> <temperature>\' \tfor setting a temperature\n";
	print "\'acer_ec ?= <reg>\' \t\tQuery register's value\n";
	print "\'acer_ec := <reg> <val>\' \tSet register's value\n";
	print "\'acer_ec +f <reg> <val>\' \tOr register's value with val (to set flags)\n";
	print "\'acer_ec -f <reg> <val>\' \tAnd register's value with ~val (to clear flags)\n";
	print "where <temp-number> is from 0-7, and <temperture> is from 0-255\n";
} elsif ($ARGV[0] eq "regs") {
	print_regs();
} elsif ($ARGV[0] eq "getled") {
	# TM2490 only (WMID)
	initialize_ioports();
	if (read_ec(0x9f)&0x01) {
		print "Mail LED on\n";
	} else {
		print "Mail LED off\n"; }
	close_ioports();
} elsif ($ARGV[0] eq "getled2") {
	# Aspire 5020 only (AMW0)
	initialize_ioports();
	if (read_ec(0x0A)&0x80) {
		print "Mail LED on\n";
	} else {
		print "Mail LED off\n"; }
	close_ioports();
} elsif ($ARGV[0] eq "getwireless") {
	# Aspire 5020 only (AMW0)
	initialize_ioports();
	if (read_ec(0x0A)&0x4) {
		print "Wireless on\n";
	} else {
		print "Wireless off\n"; }
	close_ioports();
} elsif ($ARGV[0] eq "gettouch") {
	# TM2490 only - needs testing
	initialize_ioports();
	if (read_ec(0x9e)&0x08) {
		print "touchpad disabled\n";
	} else {
		print "touchpad enabled\n"; }
	close_ioports();
} elsif ($ARGV[0] eq "?=") {
	initialize_ioports();
	my $r = hex($ARGV[1]);
	printf("REG[0x%02x] == 0x%02x\n", $r, read_ec($r));
	close_ioports();
} elsif ($ARGV[0] eq ":=") {
	initialize_ioports();
	my $r = hex($ARGV[1]);
	my $f = hex($ARGV[2]);
	my $val = read_ec($r);
	printf("REG[0x%02x] == 0x%02x\n", $r, $val);
	printf("REG[0x%02x] := 0x%02x\n", $r, $f);
        write_ec( $r, $f);
	printf("REG[0x%02x] == 0x%02x\n", $r, read_ec($r));
	close_ioports();
} elsif ($ARGV[0] eq "+f") {
	initialize_ioports();
	my $r = hex($ARGV[1]);
	my $f = hex($ARGV[2]);
	my $val = read_ec($r);
	printf("REG[0x%02x] == 0x%02x\n", $r, $val);
	printf("REG[0x%02x] := 0x%02x\n", $r, $val | $f);
        write_ec( $r, $val | $f);
	printf("REG[0x%02x] == 0x%02x\n", $r, read_ec($r));
	close_ioports();
} elsif ($ARGV[0] eq "-f") {
	initialize_ioports();
	my $r = hex($ARGV[1]);
	my $f = hex($ARGV[2]);
	my $val = read_ec($r);
	printf("REG[0x%02x] == 0x%02x\n", $r, $val);
	printf("REG[0x%02x] := 0x%02x\n", $r, $val & ~$f);
        write_ec( $r, $val & ~$f);
	printf("REG[0x%02x] == 0x%02x\n", $r, read_ec($r));
	close_ioports();
} elsif ($ARGV[0] eq "ledon") {
	# TM2490 only - needs testing
	initialize_ioports();
        send_ec (0x59, 0x92);
	close_ioports();
} elsif ($ARGV[0] eq "ledoff") {
	# TM2490 only - needs testing
	initialize_ioports();
        send_ec (0x59, 0x93);
	close_ioports();
} elsif ($ARGV[0] eq "getfanthresh") {
	initialize_ioports();
	$ii=read_ec(0xa9);
	close_ioports();
        print "Temperature threshhold: $ii (celsius)\n";
} elsif (($ARGV[0] eq "setfanthresh") && testnum($ARGV[1])) {
	write_temp(0xA9,$ARGV[1]);
} elsif ($ARGV[0] eq "setbright") {
	# Aspire 5020 only (AMW0)
	if ($ARGV[1] >= 0 && $ARGV[1] <= 15) {
		write_temp(0x83, $ARGV[1]);
	} else {
		print "second argument must be a number between 0 and 15\n";
	}
} else {
	print "wrong arguments!\n";
}
